src/lib.rs

```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
-
-impl LintPass for HumptyPass {
+impl LintPass for Pass {
     fn get_lints(&self) -> LintArray {
-        lint_array!(DROP_VIOLATION)
+        lint_array!(TEST_LINT)
     }
 
-    fn check_fn(&mut self, cx: &Context, _: visit::FnKind, _: &FnDecl, block: &Block, _: Span, _: NodeId) {
-        // TODO error if FnDecl contains protected types, unless there's
-        // the right attribute on this
-        let mut visitor = DropVisitor::new(cx, block.id);
+    fn check_fn(&mut self, cx: &Context, _: visit::FnKind, decl: &FnDecl, block: &Block, span: Span, id: NodeId) {
+        // Walk the arguments and add them to the map
+        let mut visitor = MyVisitor::new(cx, block.id);
+        for arg in decl.inputs.iter() {
+            visitor.walk_pat_and_add(&arg.pat);
+        }
+
         visitor.visit_block(block);
+
+        let attrs = cx.tcx.map.attrs(id);
+        for var in visitor.map.iter() {
+            // TODO: prettify
+            if !can_drop(cx.tcx, attrs, var.0) {
+                cx.tcx.sess.span_err(block.span, &format!("dropped var: {:?}", var));
+            }
+        }
+
     }
 }
 
-struct DropVisitor<'a : 'b, 'tcx : 'a, 'b> {
+#[derive(Clone)]
+struct MyVisitor<'a : 'b, 'tcx : 'a, 'b> {
     // Type context, with all the goodies
+    map: NodeMap<Span>, // (blockid and span for declaration)
     cx: &'b Context<'a, 'tcx>,
-    // No need to store the span, we can
-    // do a lookup on the Map if we wish
-    map: NodeMap<(NodeId, Span)>,
-    // The current block id
-    current_block: NodeId,
 }
 
-
-impl<'a, 'tcx, 'b> DropVisitor<'a, 'tcx, 'b> {
-    fn new(cx: &'b Context<'a, 'tcx>, block: NodeId) -> DropVisitor<'a, 'tcx, 'b> {
-        DropVisitor {
-            cx: cx,
-            map: FnvHashMap(),
-            current_block: block,
-        }
+impl <'a, 'tcx, 'b> MyVisitor<'a, 'tcx, 'b> {
+    fn new(cx: &'b Context<'a, 'tcx>, id: NodeId) -> Self {
+        let map = FnvHashMap();
+        let visitor = MyVisitor { cx: cx,
+                                  map: map,
+        };
+        visitor
     }
 
-    fn walk_pat_and_add(&mut self, pat: &Pat, block: NodeId) {
+    fn walk_pat_and_add(&mut self, pat: &Pat) {
         let ty = ty::pat_ty(self.cx.tcx, pat);
-        let mut protected = 0u8;
-        ty::walk_ty(ty, |t| {
+        ty::maybe_walk_ty(ty, |t| {
             match t.sty {
+                // We can ignore references, since they can at most contains
+                // references to other values, we're only interested owned
+                // values.
+                ty::ty_rptr(_, _) => return false,
                 ty::ty_enum(did, _) | ty::ty_struct(did, _) => {
-                    if ty::has_attr(self.cx.tcx, did, "drop_protection") {
-                        protected += 1;
-                        return;
+                    if ty::has_attr(self.cx.tcx, did, "drop_protect") {
+                        self.cx.tcx.sess.span_note(pat.span, "Adding drop protected type to map");
```

Comment:
Note that we can only insert one of them into the hashmap (it's keyed by nodeids, and the whole pat is a single node with no subnodes IIRC since pats themselves aren't nodes).